sdk/core/core-util/src/delay.ts

```diff
@@ -19,33 +19,71 @@ export interface DelayOptions {
   abortErrorMsg?: string;
 }
 
+/**
+ *
+ * @param inputs - The inputs for creating an abortable promise are the
+ *                 buildPromise function and the cleanupBeforeAbort function.
+ *                 buildPromise takes both the resolve and reject functions as
+ *                 parameters. cleanupBeforeAbort is called right before the
+ *                 promise is rejected.
+ * @returns a function that takes an optional DelayOptions parameter and returns
+ *          a promise that can be aborted.
+ * @internal
+ */
+export function createAbortablePromise<T>(inputs: {
+  buildPromise: (inputs: {
+    resolve: (value: T | PromiseLike<T>) => void;
+    reject: (reason?: any) => void;
+  }) => void;
+  cleanupBeforeAbort?: () => void;
+}): (options?: DelayOptions) => Promise<T> {
+  const { buildPromise, cleanupBeforeAbort } = inputs;
+  return ({ abortSignal, abortErrorMsg } = {}) =>
```

Comment:
what about having the `abortSignal` and the `abortErrorMsg` be part of the original options bag instead of returning a function here? I feel like it's a little nicer to be able to implement delay like this:

```ts
  return createAbortablePromise<void>({
    buildPromise: ({ resolve }) => {
      token = setTimeout(resolve, timeInMs);
    },
    cleanupBeforeAbort: () => clearTimeout(token),
    abortSignal,
    abortErrorMsg: abortErrorMsg ?? StandardAbortMessage,
  });
```

or if we use my other suggestions:

```ts
  return createAbortablePromise<void>(
    (resolve) => {
      token = setTimeout(resolve, timeInMs);
    }, {
      onAbortCalled: () => clearTimeout(token),
      abortSignal,
      abortErrorMsg
    }
  );
```

Reply:
Good point, the customer can create the function themselves if they need to. Addressed in https://github.com/Azure/azure-sdk-for-js/pull/24821/commits/64ea6c224b318c12ba37a1b94c45567f983e532c